### PR TITLE
Fixed ingest pipeline script issue

### PR DESCRIPTION
**Pull request by vikasvb90**
### Description
Additional validations in ingest pipeline script processor

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/opensearch/ingest/common/ScriptProcessor.java [hunk 1]
```diff
@@ -102,8 +102,11 @@ public IngestDocument execute(IngestDocument document) {
         } else {
             ingestScript = precompiledIngestScript;
         }
-        ingestScript.execute(document.getSourceAndMetadata());
-        CollectionUtils.ensureNoSelfReferences(document.getSourceAndMetadata(), "ingest script");
+        IngestDocument mutableDocument = new IngestDocument(document);
+        ingestScript.execute(mutableDocument.getSourceAndMetadata());
+        CollectionUtils.ensureNoSelfReferences(mutableDocument.getSourceAndMetadata(), "ingest script");
+        document.getSourceAndMetadata().clear();
+        document.getSourceAndMetadata().putAll(mutableDocument.getSourceAndMetadata());
```
Comment (andrross) [inline modules/ingest-common/src/main/java/org/opensearch/ingest/common/ScriptProcessor.java:109]: What exactly is happening here? A naive reading of this code looks like `document` is being mutated but `mutableDocument` is not being mutated.

File: server/src/main/java/org/opensearch/ingest/IngestDocument.java [hunk 2]
```diff
@@ -752,6 +753,7 @@ public Map<String, Object> getSourceAndMetadata() {
 
     @SuppressWarnings("unchecked")
     public static <K, V> Map<K, V> deepCopyMap(Map<K, V> source) {
+        CollectionUtils.ensureNoSelfReferences(source, "IngestDocument: Self reference present in object.");
```
Comment (andrross) [inline server/src/main/java/org/opensearch/ingest/IngestDocument.java:756]: Can we add some unit tests here? I'm guessing without this check this code could result in a stack overflow?

Comment (vikasvb90) [reply server/src/main/java/org/opensearch/ingest/IngestDocument.java:756]: done

File: modules/ingest-common/src/main/java/org/opensearch/ingest/common/ScriptProcessor.java (hunk 1 above)
Comment (vikasvb90) [reply modules/ingest-common/src/main/java/org/opensearch/ingest/common/ScriptProcessor.java:109]: `mutableDocument` is being mutated and then after validation changes are being applied to original doc.

**Review by shwetathareja — APPROVED** [review]
